### DISCUSSION TITLE
Android Orientation

### DISCUSSION THREAD

**Discussion by salisuabubakar** [grahammendick/navigation#690] [Q&A] [answered]
Hi @grahammendick,
Is there a way one can set orientation based on a scene ?  I would like to be able to do that on android and ios. 

Thanks

**Comment by grahammendick** [comment]
That isn't part of the Navigation router. Is there a 3rd party library you can use?

**Reply by grahammendick** [reply]
Everything ok? Have you found a 3rd party library?

**Reply by salisuabubakar** [reply]
Hi ,
I couldn’t get the 3rd party library, am sure most people would love to see this on The navigation router.

Thanks

**Reply by grahammendick** [reply]
I try to only bring the navigation primitives into the Navigation router. But orientation might be one that I've missed. Thanks for bringing it to my attention.

**Reply by grahammendick** [reply]
Early thoughts, I'm planning on adding an 'orientation' prop to `Scene`
```jsx
<Scene stateKey="movie" orientation="landscape">
```
It will also accept a function so you can set the orientation based on navigation data
```jsx
<Scene stateKey="movie" orientation={({ fullScreen }) => fullScreen ? 'landscape' : undefined}>
```
So you can also change the orientation from a button press within the movie scene
```jsx
<Button onPress={() => stateNavigator.refresh({ fullScreen: true })} >
```

How does that sound?

**Reply by salisuabubakar** [reply]
Wow that would be awesome. Am really looking forward to  that. Thanks for your time.

**Comment by salisuabubakar** [comment]
HI, Any Specific time to get this done. Thanks for your time.

**Reply by grahammendick** [reply]
Would you like to work on it? I'll help you as much as you need

**Reply by salisuabubakar** [reply]
Hi, 
Thanks for the response. Am coming from Web development background. Can you show me what to do and guide me. Thanks

**Reply by grahammendick** [reply]
Nice one. Thanks for giving it a go. Have you got Android Studio and Xcode set up? You'll need these because you'll be editing some native code. I'll help you with all this, but I want to make sure you're environment is ready

**Reply by salisuabubakar** [reply]
Yes am ready. I've got Android Studio and Xcode set up.

**Reply by grahammendick** [reply]
Theh rough idea is that you'll add an `orientation` prop to the `Scene` component. React Native will send this prop to the native side. You'll get this prop in Java (Android) and Objective-C (iOS) and then set the native orientation to match.

So a good place for you to start is to work out how to set the orientation in Java and Objective-C. While you're doing that I'll write a high level spec of the changes. Does that sound like a good plan?

**Reply by grahammendick** [reply]
I should be free to start on this soon. If you still want to work on it let me know otherwise I'm gonna assume you're happy for me to pick this up

**Reply by salisuabubakar** [reply]
Hi, thanks for your response. Am coming from the web development environment. Am learning how to get the orientation in Android and IOS. If you can take it up i will be very grateful. Thanks

**Reply by grahammendick** [reply]
I've started working on this, you can [follow along if you want](https://github.com/grahammendick/navigation/compare/scene-orientation)

**Reply by grahammendick** [reply]
I've got landscape working with Android but I'm struggling with iOS. I get the feeling that you're only meant to change the orientation for [full screen modals on iOS](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621438-preferredinterfaceorientationfor?language=objc#:~:text=the%20view%20controller%20full%20screen). I don't have an iPhone so I can't look at any real life examples. Do you know?

**Reply by grahammendick** [reply]
I'm convinced that's right because [React Native's Modal has a supportedOrientations prop](https://reactnative.dev/docs/modal#supportedorientations-ios) that rotates the screen when set to 'landscape'. And if you watch [Apple's UIKit video at 23:20](https://developer.apple.com/videos/play/wwdc2022/10068/) it basically says it's only supported in full screen view controllers.

Is that why you only asked for Android?

**Reply by salisuabubakar** [reply]
Yes. Can you please do the update for the android you got working?. Thanks

**Reply by grahammendick** [reply]
Can you explain a bit more about why you want to change the orientation, please?

**Comment by salisuabubakar** [comment]
Am using it in my app that I migrated from Wix RNN to The Navigation
Router. It’s a video streaming app and when user click to watch any movie
the orientation should change from portrait to landscape.

Thank you.

On Sunday, April 30, 2023, Graham Mendick ***@***.***> wrote:

> Can you explain a bit more about why you want to change the orientation,
> please?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/grahammendick/navigation/discussions/690#discussioncomment-5765201>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABURJHF6GO3BL37SLLDIFGLXDYPSVANCNFSM6AAAAAAVZWFDMI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
*SALISU ABUBAKAR*
Fullstack Developer / Consultant
*Mobile: 233257335982*
*Email  : ***@***.***

**Reply by grahammendick** [reply]
Thanks for the info. What happens on iOS when the user clicks to watch a movie? Are you opening the movie in a Modal and setting the [supportedOrientations prop](https://reactnative.dev/docs/modal#supportedorientations-ios) to 'landscape'?
What do apps like Netflix do on an iPhone?

**Reply by salisuabubakar** [reply]
Hi,
Sorry for the delay.  When the user navigate to the video page i want the default orientation to change from Portrait to landscape. Am using react native webview to play the video. Sorry once again for the delay. Thanks

**Reply by grahammendick** [reply]
How are you going to handle orientation change on iOS?

**Reply by salisuabubakar** [reply]
Hi, am not sure how to do that. Is it possible you add screen orientation for both android and ios so that we can change the screen orientation for each screen. I came from the web background. Am not good with java/kotlin or swift. Thanks

**Reply by grahammendick** [reply]
I'm pretty sure you're not meant to change the screen orientation on iOS anymore. [Apple's UIKit video at 23:20](https://developer.apple.com/videos/play/wwdc2022/10068/) basically says it's only supported in full screen view controllers. So only in a Modal with `presentationStyle` set to `fullScreen`. Then you can set the [supportedOrientations prop to landscape](https://reactnative.dev/docs/modal#supportedorientations-ios).

**Comment by salisuabubakar** [comment]
Hi, I think I understood what you mean. I'll use modal on ios. Can you help
me get the android done? Thank you

On Tue, Jun 13, 2023 at 1:02 PM Graham Mendick ***@***.***>
wrote:

> I'm pretty sure you're not meant to change the screen orientation on iOS
> anymore. Apple's UIKit video at 23:20
> <https://developer.apple.com/videos/play/wwdc2022/10068/> basically says
> it's only supported in full screen view controllers. So only in a Modal
> with presentationStyle set to fullScreen. Then you can set the supportedOrientations
> prop to landscape
> <https://reactnative.dev/docs/modal#supportedorientations-ios>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/grahammendick/navigation/discussions/690#discussioncomment-6163453>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABURJHDQHHJPNQ5AZFYHTMTXLBQETANCNFSM6AAAAAAVZWFDMI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***
> com>
>


-- 
*SALISU ABUBAKAR*
Fullstack Developer / Consultant
*Mobile: 233257335982*
*Email  : ***@***.***

**Reply by grahammendick** [reply]
Yea, I should be able to pick up where I left off [in the scene-orientation branch](https://github.com/grahammendick/navigation/compare/scene-orientation). You happy with the following?
```jsx
<Scene stateKey="movie" landscape={true}>
```
It will also accept a function so you can set to landscape based on navigation data, for example, when pressing a full screen button
```jsx
<Scene stateKey="movie" landscape={({ fullScreen }) => fullScreen}>
```

**Reply by salisuabubakar** [reply]
Yes am very happy. I really appreciate. let me know when i can help you test. Thank you.

**Reply by grahammendick** [reply]
Oh great, thanks for offering to help. I'll let you know when it's ready for testing

**Reply by salisuabubakar** [reply]
Thanks for this wonderful library. The Navigation Router has given me absolute confidence. Thanks soo much. Really appreciate.

**Reply by grahammendick** [reply]
Ok, it's ready for testing. 

1. Clone the navigation repository and check out the `scene-orientation` branch
2. Run `npm run package` from the top level folder
3. Copy the `navigation-react-native` folder from the generated `build/npm` folder into the `node_modules` of your app
4. Set the `landscape` prop to true on the `Scene` you want to be in landscape

Any questions please let me know.

I did that to the Timeline Scene of the twitter sample and it looks like the vid below.
```jsx
<Scene stateKey="timeline" landscape><Timeline /></Scene>
```

https://github.com/grahammendick/navigation/assets/1761227/2b8f91e3-06d4-46f4-a73e-f8c1f3694df7

**Reply by salisuabubakar** [reply]
Yes i have tested and its working just like your video. Thanks. How soon are you going to merge or update the library. Thanks

**Reply by grahammendick** [reply, marked as answer]
Just merged in landscape support in #706. Thanks for raising the issue. I'll thank you in the release notes, too

**Reply by grahammendick** [reply]
Just published navigation-react-native 9.3.0 with landscape support. I [thanked you in the release notes](https://github.com/grahammendick/navigation/releases/tag/v9.3.0-NavigationReactNative)
